akka-actor/src/main/scala/akka/io/TcpConnection.scala

```diff
@@ -389,6 +398,12 @@ private[io] abstract class TcpConnection(val tcp: TcpExt, val channel: SocketCha
           if (data eq remainingData) this
           else new PendingBufferWrite(commander, data, ack, buffer, tail) // copy with updated remainingData
 
+        } else if (data.nonEmpty && buffer.isReadOnly) {
+          val head = data.headByteString
+          val remaining = data.drop(head.length)
+
+          //this will always start with the direct buffer, or will change over to atmost one
+          PendingBufferWrite(commander, remaining, ack, tail).doWrite(info)
```

Comment:
This is technically open recursion, it might be possible to construct a pathological write request that blows the stack.